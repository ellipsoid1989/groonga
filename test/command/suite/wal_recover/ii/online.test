# _database_close/_database_reopen can't use with HTTP because requests via HTTP are always
# processed by workers. Workers never use primary WAL role.
#@require-interface stdio

#@require-feature message_pack

#$GROONGA_ALLOW_DATABASE_REOPEN=yes
#$GRN_WAL_ROLE=primary

#@disable-logging
cache_limit --max 0
#@enable-logging

table_create Memos TABLE_NO_KEY
column_create Memos content COLUMN_SCALAR Text

table_create Terms TABLE_PAT_KEY ShortText \
  --default_tokenizer TokenNgram \
  --normalizer NormalizerNFKC130
column_create Terms memos_content COLUMN_INDEX|WITH_POSITION Memos content

io_flush
#@copy-path #{db_path}.0000103 #{db_path}.0000103.initial
#@copy-path #{db_path}.0000103.c #{db_path}.0000103.c.initial

load --table Memos
[
{"content": "hello world"},
{"content": "good-by world"}
]

check --obj Terms.memos_content

select Memos --query content:@hello

#@copy-path #{db_path}.0000103.wal #{db_path}.0000103.wal.keep
_database_close

#@copy-path #{db_path}.0000103.initial #{db_path}.0000103
#@copy-path #{db_path}.0000103.c.initial #{db_path}.0000103.c
#@copy-path #{db_path}.0000103.wal.keep #{db_path}.0000103.wal
#@add-important-log-levels notice
_database_reopen
#@remove-important-log-levels notice

select Memos --query content:@hello

check --obj Terms.memos_content
