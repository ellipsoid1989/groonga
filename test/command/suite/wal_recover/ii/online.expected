table_create Memos TABLE_NO_KEY
[[0,0.0,0.0],true]
column_create Memos content COLUMN_SCALAR Text
[[0,0.0,0.0],true]
table_create Terms TABLE_PAT_KEY ShortText   --default_tokenizer TokenNgram   --normalizer NormalizerNFKC130
[[0,0.0,0.0],true]
column_create Terms memos_content COLUMN_INDEX|WITH_POSITION Memos content
[[0,0.0,0.0],true]
io_flush
[[0,0.0,0.0],true]
load --table Memos
[
{"content": "hello world"},
{"content": "good-by world"}
]
[[0,0.0,0.0],2]
check --obj Terms.memos_content
[
  [
    0,
    0.0,
    0.0
  ],
  [
    {
      "flags": "00008202",
      "max sid": 1,
      "number of garbage segments": 0,
      "number of array segments": 1,
      "max id of array segment": 1,
      "number of buffer segments": 1,
      "max id of buffer segment": 1,
      "max id of physical segment in use": 1,
      "number of unmanaged segments": 0,
      "total chunk size": 0,
      "max id of chunk segments in use": 0,
      "number of garbage chunk": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ]
    },
    {
      "buffer id": 0,
      "void chunk size": 0,
      "buffer term": [
        [
          "world",
          2,
          2,
          2,
          0
        ]
      ],
      "buffer free": 262096,
      "size in buffer": 2,
      "nterms": 1,
      "nterms with chunk": 0
    }
  ]
]
select Memos --query content:@hello
[[0,0.0,0.0],[[[1],[["_id","UInt32"],["content","Text"]],[1,"hello world"]]]]
_database_close
[[0,0.0,0.0],true]
_database_reopen
[[0,0.0,0.0],true]
#|e| [ii][wal][recover][Terms.memos_content] not implemented: path:<db/db.0000103>
#|n| [db][wal][recover] rebuild broken column: <Terms.memos_content>(259)
#|n| DDL:260:column_create Terms #recovering#memos_content
#|n| spec:260:update:Terms.#recovering#memos_content:72(column:index):256(Memos)
#|n| DDL:260:set_source Terms.#recovering#memos_content Memos.content
#|n| spec:257:update:Memos.content:65(column:var_size):15(Text)
#|n| DDL:2147483649:set_normalizers NormalizerNFKC130
#|n| DDL:2147483649:set_normalizers NormalizerNFKC130
#|n| spec:260:update:Terms.#recovering#memos_content:72(column:index):256(Memos)
#|n| DDL:259:obj_remove Terms.memos_content
#|n| spec:257:update:Memos.content:65(column:var_size):15(Text)
#|n| spec:259:remove:Terms.memos_content:72(column:index)
#|n| [db][wal][recover] succeeded to rebuild broken column: <Terms.memos_content>(260)
select Memos --query content:@hello
[[0,0.0,0.0],[[[1],[["_id","UInt32"],["content","Text"]],[1,"hello world"]]]]
check --obj Terms.memos_content
[
  [
    0,
    0.0,
    0.0
  ],
  [
    {
      "flags": "00008202",
      "max sid": 0,
      "number of garbage segments": 0,
      "number of array segments": 1,
      "max id of array segment": 1,
      "number of buffer segments": 1,
      "max id of buffer segment": 1,
      "max id of physical segment in use": 1,
      "number of unmanaged segments": 0,
      "total chunk size": 1048576,
      "max id of chunk segments in use": 1,
      "number of garbage chunk": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ]
    },
    {
      "buffer id": 0,
      "chunk size": 1048576,
      "buffer term": [
        [
          "world",
          2,
          2,
          0,
          7,
          2,
          2
        ]
      ],
      "buffer free": 262112,
      "size in buffer": 0,
      "nterms": 1,
      "nterms with chunk": 1
    }
  ]
]
